pkg/controller/atlasdeployment/atlasdeployment_controller.go

```diff
@@ -153,53 +153,67 @@ func (r *AtlasDeploymentReconciler) Reconcile(context context.Context, req ctrl.
 	// Allow users to specify M0/M2/M5 deployments without providing TENANT for Normal and Serverless deployments
 	r.verifyNonTenantCase(deployment)
 
-	if result := r.checkDeploymentIsManaged(workflowCtx, context, log, project, deployment); !result.IsOk() {
-		return result.ReconcileResult(), nil
+	// convertedDeployment is either serverless or advanced, deployment must be kept unchanged
+	// convertedDeployment is always a separate copy, to avoid changes on it to go back to k8s
+	convertedDeployment := deployment.DeepCopy()
+	if deployment.IsLegacyDeployment() {
+		if err := ConvertLegacyDeployment(&convertedDeployment.Spec); err != nil {
+			result = workflow.Terminate(workflow.Internal, err.Error())
+			log.Errorw("failed to convert legacy deployment", "error", err)
+			return result.ReconcileResult(), nil
+		}
+		convertedDeployment.Spec.DeploymentSpec = nil
 	}
 
-	deletionRequest, result := r.handleDeletion(workflowCtx, context, log, prevResult, project, deployment)
-	if deletionRequest {
+	if result := r.checkDeploymentIsManaged(workflowCtx, context, log, project, convertedDeployment); !result.IsOk() {
 		return result.ReconcileResult(), nil
 	}
 
-	err = customresource.ApplyLastConfigApplied(context, deployment, r.Client)
-	if err != nil {
-		result = workflow.Terminate(workflow.Internal, err.Error())
-		workflowCtx.SetConditionFromResult(status.DeploymentReadyType, result)
-		log.Error(result.GetMessage())
-
+	deletionRequest, result := r.handleDeletion(workflowCtx, context, log, prevResult, project, deployment)
+	if deletionRequest {
 		return result.ReconcileResult(), nil
 	}
 
-	if deployment.IsLegacyDeployment() {
-		if err := ConvertLegacyDeployment(&deployment.Spec); err != nil {
-			result = workflow.Terminate(workflow.Internal, err.Error())
-			log.Errorw("failed to convert legacy deployment", "error", err)
-			return result.ReconcileResult(), nil
-		}
-		deployment.Spec.DeploymentSpec = nil
-	}
-
-	if err := uniqueKey(&deployment.Spec); err != nil {
+	if err := uniqueKey(&convertedDeployment.Spec); err != nil {
 		log.Errorw("failed to validate tags", "error", err)
 		result := workflow.Terminate(workflow.Internal, err.Error())
 		workflowCtx.SetConditionFromResult(status.DeploymentReadyType, result)
 		return result.ReconcileResult(), nil
 	}
 
-	handleDeployment := r.selectDeploymentHandler(deployment)
-	if result, _ := handleDeployment(workflowCtx, project, deployment, req); !result.IsOk() {
+	handleDeployment := r.selectDeploymentHandler(convertedDeployment)
+	if result, _ := handleDeployment(context, workflowCtx, project, convertedDeployment, req); !result.IsOk() {
 		workflowCtx.SetConditionFromResult(status.DeploymentReadyType, result)
-		return result.ReconcileResult(), nil
+		return r.registerConfigAndReturn(workflowCtx, context, log, deployment, result), nil
```

Comment:
Setting the config for non-successful operations may cause trouble in the next reconciliation once we will be comparing to a "wrong" configuration. I think we should separate the intermediate state "in progress"(creating/updating) from actual errors when saving the config. The same applies to serverless.

Reply:
It is separated:
```go
func (r *AtlasDeploymentReconciler) registerConfigAndReturn(
...
	if result.IsOk() || result.IsInProgress() {
		err := customresource.ApplyLastConfigApplied(context, deployment, r.Client)
...
```

Errors will not save the config, OK and in progress will.